### PR TITLE
BEAM-307: Upgrade/Test to Kafka 0.10

### PR DESCRIPTION
**Pull request by XuMingmin**
Run *KafkaIOTest.java* with `mvn test -Dkafka.clients.version=0.10.1.1`, or `mvn test -Dkafka.clients.version=0.9.0.1` for either Kafka client version.

it can work with Kafka client 0.9 and 0.10 as:
1. SpEL can find this function, either input is List or Collection;
2. List extends Collection, so super.assign() could find either assign(List) or assign(Collection).

### CODE REVIEW

**Comment by davorbonaci** [comment]
R: @davorbonaci

CC: @rangadi @dhalperi @jasonkuster @ssisk

File: sdks/java/io/kafka/pom.xml [hunk 1]
```diff
@@ -29,6 +29,10 @@
   <name>Apache Beam :: SDKs :: Java :: IO :: Kafka</name>
   <description>Library to read Kafka topics.</description>
 
+  <properties>
+    <kafka.clients.version>0.10.1.1</kafka.clients.version>
```
Comment (rangadi) [inline sdks/java/io/kafka/pom.xml]: Can we keep the default to 0.9.0.1? Looks like in Kafka, older clients can connect to newer Kafka servers, but not the other way around.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 2]
```diff
@@ -100,7 +100,7 @@
  * An unbounded source and a sink for <a href="http://kafka.apache.org/">Kafka</a> topics.
  * Kafka version 0.9 and 0.10 are supported. If you need a specific version of Kafka
  * client(e.g. 0.9 for 0.9 servers, or 0.10 for security features), specify explicit
- * kafka-client dependency.
+ * kafka-client dependency with Maven property ${kafka.clients.version}.
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: I think we should remove this update. Uses are not building beam or beam modules by themselves, rather are just using published mvn artifacts. This property is not irrelevant to them.

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java [hunk 3]
```diff
@@ -143,7 +145,10 @@
     final MockConsumer<byte[], byte[]> consumer =
         new MockConsumer<byte[], byte[]>(offsetResetStrategy) {
           // override assign() in order to set offset limits & to save assigned partitions.
-          @Override
+          //remove keyword '@Override' here, it can work with Kafka client 0.9 and 0.10 as:
+          //1. SpEL can find this function, either input is List or Collection;
+          //2. List extends Collection, so super.assign() could find either assign(List)
```
Comment (rangadi) [inline sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java]: Makes sense. Thanks for the comment.

File: sdks/java/io/kafka/pom.xml (hunk 1 above)
Comment (XuMingmin) [inline sdks/java/io/kafka/pom.xml]: my bad, it should be kept as 0.9.0.1

**Comment by rangadi** [comment]
👍 LGTM. Thanks Xu.

**Comment by XuMingmin** [comment]
@davorbonaci , more details about 'taking this on'?

**Comment by davorbonaci** [comment]
This PR is making it possible to run against multiple versions of Kafka with a single flag. The CI infrastructure is, however, using one, specific version. So, we won't know if/when something breaks here. It would be great to automate testing against multiple versions.

This task would be non-trivial. It can be a new module that changes the dependency. Or, it can be another test invocation in the same module that overrides the version for the second invocation. (Other options available too.) All this may or may not involve spinning multiple-versions of Kafka clusters via, say, Kubernates.

Overall, this is a complex infrastructure work -- it is totally fine if this is not something you are personally interested in. We'll get there eventually ;-)

**Comment by rangadi** [comment]
Simplest is extra test invocation in one one of the builds (ci, jenkins, or nightly): 
`mvn test -Dkafka.clients.version=0.10.1.1 -Dtest=Kafka\* -DfailIfNoTests=false`.

e.g. we could add this to `/.travis.yml`. This ensures the code itself works ok.

**Comment by davorbonaci** [comment]
Indeed -- but, that approach won't scale for our project-wide needs, and goes against many of the goals we have set out for our engineering system (e.g., easy for contributors to run the tests, no complicated commands to reproduce what the CI is running, interop with plugins, etc.). So, I'd like to go beyond this.

**Comment by XuMingmin** [comment]
I'll look into this after timestamp support.
Mostly I'll look for a test plugin in pom, so the impact is limited to this module only.

**Pull request closed without merging**